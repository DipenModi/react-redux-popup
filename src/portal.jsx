import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { CSSTransitionGroup } from 'react-transition-group';
import collection, { TYPE_MODAL, TYPE_POPUP } from 'rrp/popup-collection';

export const popupSelector = state => state.popup;

function renderPopups(popupType, props) {
    return collection
        .filter(popup => popup[0] === popupType && props[popup[2].id])
        // eslint-disable-next-line no-unused-vars
        .map(([type, Popup, popupProps]) => <Popup key={popupProps.id} {...popupProps} />);
}

export const Portal = props => (
    <div>
        <CSSTransitionGroup
            transitionName={props.modalTransitionName}
            transitionEnterTimeout={props.modalTransitionEnterTimeout}
            transitionLeaveTimeout={props.modalTransitionLeaveTimeout}
        >
            {renderPopups(TYPE_MODAL, props)}
        </CSSTransitionGroup>
        <CSSTransitionGroup
            transitionName={props.popupTransitionName}
            transitionEnterTimeout={props.popupTransitionEnterTimeout}
            transitionLeaveTimeout={props.popupTransitionLeaveTimeout}
        >
            {renderPopups(TYPE_POPUP, props)}
        </CSSTransitionGroup>
    </div>
);

Portal.displayName = 'Portal';

Portal.propTypes = {
    modalTransitionName: PropTypes.string.isRequired,
    modalTransitionEnterTimeout: PropTypes.number.isRequired,
    modalTransitionLeaveTimeout: PropTypes.number.isRequired,
    popupTransitionName: PropTypes.string.isRequired,
    popupTransitionEnterTimeout: PropTypes.number.isRequired,
    popupTransitionLeaveTimeout: PropTypes.number.isRequired
};

Portal.defaultProps = {
    modalTransitionName: 'modal',
    modalTransitionEnterTimeout: 0,
    modalTransitionLeaveTimeout: 0,
    popupTransitionName: 'popup',
    popupTransitionEnterTimeout: 0,
    popupTransitionLeaveTimeout: 0
};

export default connect(popupSelector)(Portal);
