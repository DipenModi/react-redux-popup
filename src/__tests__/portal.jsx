import { mount, shallow } from 'enzyme';
// import configureStore from 'redux-mock-store';
// import { Provider } from 'react-redux';
import { Portal, getPortalElement, setPortalElement } from 'rrp/portal';

describe('portal', () => {
    it('should render the component', () => {
        const wrapper = shallow(<Portal portalInitialized={() => true} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should initialize the portal element on mount', () => {
        const portalInitialized = jest.fn();
        const wrapper = mount(<Portal portalInitialized={portalInitialized} />);
        expect(wrapper).toMatchSnapshot();
        expect(portalInitialized).toHaveBeenCalledTimes(1);
    });

    it('should not update on prop changes', () => {
        const spy = jest.spyOn(Portal.prototype, 'shouldComponentUpdate');

        const portalInitialized = jest.fn();
        const wrapper = shallow(<Portal portalInitialized={portalInitialized} />);
        wrapper.setProps({ portalInitialized: () => true });

        expect(spy).toHaveBeenCalledTimes(1);

        spy.mockReset();
        spy.mockRestore();
    });

    it('should set and get portal element', () => {
        const elem = 'some elem';
        setPortalElement(elem);
        expect(getPortalElement()).toBe(elem);
    });
});
