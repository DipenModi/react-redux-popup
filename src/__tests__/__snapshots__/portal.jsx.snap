// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portal should not render the model when id is not present 1`] = `
<div>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="modal"
  />
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="popup"
  />
</div>
`;

exports[`portal should not render the popup when id is not present 1`] = `
<div>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="modal"
  />
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="popup"
  />
</div>
`;

exports[`portal should render the component 1`] = `
<div>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="modal"
  />
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="popup"
  />
</div>
`;

exports[`portal should render the component with modal type 1`] = `
<div>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="modal"
  >
    <TestComponent
      id="modal1"
      prop1="value1"
    />
  </CSSTransitionGroup>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="popup"
  />
</div>
`;

exports[`portal should render the component with popup type 1`] = `
<div>
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="modal"
  />
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={0}
    transitionName="popup"
  >
    <TestComponent
      id="popup1"
      prop1="value1"
    />
  </CSSTransitionGroup>
</div>
`;
